nirtorch/graph_fx.py

```diff
@@ -0,0 +1,207 @@
+        Dict[torch.nn.Module, Callable[[torch.nn.Module], nir.NIRNode]]
+    ] = None,
+) -> nir.NIRGraph:
+    """
+    Traces a PyTorch module and converts it to a NIR graph using the specified module map.
+
+    Args:
+        module (torch.nn.Module): The module of interest
+        module_map (Dict[torch.nn.Module, Callable[[torch.nn.Module], nir.NIRNode]]): A dictionary that maps
+            a given module type to a function that can convert the model to an NIRNode type
+        default_dict (Optional[Dict[torch.nn.Module, Callable[[torch.nn.Module], nir.NIRNode]]]): An optional dictionary that maps
+            a given module type to a function that can convert the model to an NIRNode type. This dictionary is merged
+            with the module_map dictionary.
+    """
+    # Merge the default dictionary, if it exists
+    if default_dict is not None:
+        module_map = module_map | default_dict
+
+    # Cover the edge case that the incoming module is a leaf node
+    if module.__class__ in module_map:
+        return module_map[module.__class__](module)
+
+    # Trace the graph
+    tracer = NIRTorchTracer(module_map.keys())
+    traced = tracer.trace(module)
+
+    if len(traced.nodes) == 2 and len(list(tracer.root.children())) == 0:
+        raise ValueError(
+            "The module is a leaf node, but does not appear in the module map. We cannot trace it further"
+        )
+
+    graph_module = GraphModule(tracer.root, traced)
+
+    # Create NIR nodes
+    nodes = {}
+    edges = []
+    ignored_nodes = set()
+    skipped_nodes = set()
+
+    def _find_users(node: Node) -> Set[Node]:
+        """
+        Finds all the users (outputs) of a given node, recursively if the node is registered as a skipped node
+        """
+        nodes = set()
+        for user in node.users:
+            if user in ignored_nodes:
+                continue
+            elif user in skipped_nodes:
+                nodes |= _find_users(user)
+            else:
+                nodes.add(user)
+        return nodes
+
+    def _find_inputs(node: Node) -> Set[Node]:
+        """
+        Finds all the inputs (inputs) of a given node, recursively if the node is registered as a skipped node
+        """
+        nodes = set()
+        for in_node in node.all_input_nodes:
+            if in_node in ignored_nodes:
+                continue
+            elif in_node in skipped_nodes:
+                nodes |= _find_inputs(in_node)
+            else:
+                nodes.add(in_node)
+        return nodes
+
+    for node in traced.nodes:
+        # Add Node
+        if node.op == "placeholder":
+            if node.target == "input" or node.prev.op == "root":
+                nodes[str(node.name)] = nir.Input(np.array([1]))
+            else:
+                ignored_nodes.add(node)
+                continue
+        elif node.op == "output":
+            nodes[str(node.name)] = nir.Output(np.array([1]))
+        elif node.op == "call_function":
```

Comment:
Should we add a check for other non-allowed functions (multiplies or divisions) and add an exception?